add conversion from SAM to BAM

jonathanBieler opened the pull request:
## Types of changes

This PR implements the following changes:
_(Please tick any or all of the following that are applicable)_

* [x] :sparkles: New feature (A non-breaking change which adds functionality).
* [ ] :bug: Bug fix (A non-breaking change, which fixes an issue).
* [ ] :boom: Breaking change (fix or feature that would cause existing functionality to change).

## :clipboard: Additional detail

This adds a method to `convert` to convert a SAM record to a BAM record, it will allow to use BWA as a library, create reads in-silico, etc. I'm testing it extensively so it should mostly be correct, but there's still some corner cases I think. I'll add some documentation later. Fix #61

```julia
sam = "seq1\t81\tPhiX\t1051\t60\t70M\t=\t1821\t702\tTCTTGGCTTCCTTGCTGGTCAGATTGGTCGTCTTATTACCATTTCAACTACTCCGGTTATCGCTGGCGAC\t*\tNM:i:0\tMD:Z:70\tMC:Z:70M\tAS:i:70\tXS:i:0"
record = XAM.SAM.Record(sam)

bam_record = convert(BAM.Record, record)
```

## :ballot_box_with_check: Checklist

- [x] :art: The changes implemented is consistent with the [julia style guide](https://docs.julialang.org/en/stable/manual/style-guide/).
- [ ] :blue_book: I have updated and added relevant docstrings, in a manner consistent with the [documentation styleguide](https://docs.julialang.org/en/stable/manual/documentation/).
- [ ] :blue_book: I have added or updated relevant user and developer manuals/documentation in `docs/src/`.
- [x] :ok: There are unit tests that cover the code changes I have made.
- [x] :ok: The unit tests cover my code changes AND they pass.
- [ ] :pencil: I have added an entry to the `[UNRELEASED]` section of the manually curated `CHANGELOG.md` file for this repository.
- [x] :ok: All changes should be compatible with the latest stable version of Julia.
- [ ] :thought_balloon: I have commented liberally for any complex pieces of internal code.

kescobo reviewed (approved):
This looks good to me on a first pass, and I really appreciate all the tests! I'm not so familiar with these record types, so I definitely want someone else to take a look, but a few other things to consider in addition to my in-line comments:

1. Do any of these functions make sense to be exported? Eg parsing and encoding CIGAR seem like they could have broader utility.
2. Does this really need to go in its own module?

jonathanBieler commented on src/convert.jl:
```diff
@@ -0,0 +1,229 @@
+        '0' => 0x0a
+    )
+
+    function parse_cigar(record::SAM.Record)
```
I haven't found anything, there's just a `cigar` method that returns the cigar string. It would be good to add a `cigar_rle` method for SAM, but I went for the minimum thing I needed here.

jonathanBieler commented on src/convert.jl:
```diff
@@ -0,0 +1,229 @@
+        out
+    end
+
+    function convert(::Type{BAM.Record}, record::SAM.Record; header = nothing)
```
I import it, but maybe better to go with Base.

kescobo replied:
That's fine - I marginally prefer to do explicit `Base.` access, but if there are any other extended methods from `Base` in this package, stick to whatever convention is being used, I think

jelber2 commented:
Super cool, I just used this added functionality @jonathanBieler ! see https://gist.github.com/jelber2/47129820373474a768dacabc0e686fdc

kescobo commented:
Great! Would be nice to get it merged in so it's a bit more accessible. @jakobnissen or @CiaranOMara - have you had a chance to take a looks?

jelber2 commented on src/convert.jl:
```diff
@@ -0,0 +1,229 @@
+        bam_record.mapq = SAM.mappingquality(record)
+        bam_record.bin = bin
+        bam_record.n_cigar_op = n_cigar_op
+        bam_record.flag = SAM.flag(record)
```
This should be .flags and not .flag twice.

jonathanBieler replied:
Done.

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+
+using BioSequences, BioAlignments
+
+global  const op_to_code = Dict(
```
No need to put global here - also, usually, global constants are upper case
```suggestion
const OP_TO_CODE = Dict(
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+While all single (i.e., non-array) integer types are stored in SAM as 'i', in BAM any of 'cCsSiI' 
+may be used together with the correspondingly-sized binary integer value, chosen according to the field value's magnitude
+"""
+global const type_to_aux_int_fields = Dict(
```
```suggestion
const TYPE_TO_AUX_INT_FIELDS = Dict(
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    !SAM.hassequence(record) && return UInt8[]
+
+    Nsam = SAM.seqlength(record)
+    Nbam = iseven(Nsam) ? div(Nsam,2) : div(Nsam+1,2)
```
```suggestion
    Nbam = cld(Nsam, 2)
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+using ..SAM
+using ..BAM
+
+using BioSequences, BioAlignments
```
I recommend you use explicit imports here: Figure out exactly what names you are using an import them.
So, instead of `using BioSequences`, have: `using BioSequences: encode, DNAAlphabet`, etc.
This makes it easier to maintain, since we then know exactly what is needed from the two dependencies.

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+
+    for (i,c) in enumerate(cigar)
+        if c ∈ cigar_ops
+            L = parse(Int, cigar[start:i-1])
```
Avoid allocating a string for this
```suggestion
            L = parse(Int, view(cigar, start:i-1))
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+end
+
+function Base.convert(::Type{BAM.Record}, record::SAM.Record; header = nothing)
+
+    bam_record = BAM.Record()
+    set_fixed_length_fields!(bam_record, record, header)
+    set_variable_length_fields!(bam_record, record)
+
+    bam_record
```
```suggestion
    bam_record = BAM.Record()
    set_fixed_length_fields!(bam_record, record, header)
    set_variable_length_fields!(bam_record, record)
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    bam_record
+end
+
+function refname_to_refid(refname::String, header)
+    isnothing(header) && return -1
+    refname == "*" && return -1
+    
+    refs = findall(header, "SQ")
+    for (i,ref) in enumerate(refs)
+        ref["SN"] == refname && return i-1
+    end
+    error("Couldn't find reference $(refname) in header.")
```
This function is highly inefficient, since it needs to scan every single header for each record. This implies that perhaps there needs to be some kind of API to convert records from SAM to BAM in bulk, where the header is parsed only once, and then it builds a `Dict{String, Int}` to look up the refid from the refname.
Could you add a type annotation to `header`? Just for readability.

jonathanBieler replied:
Yeah, maybe SAM.Header could be extended to be able do that (it seems its "responsibility" to handle refnames/ids).

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+
+function parse_cigar(record::SAM.Record)
+    
+    cigar_ops = ('M','I','D','N','S','H','P','=','X')
```
You could have this be a tuple of bytes instead, and then loop over the `codeunits(cigar)`, for speed.

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+function set_fixed_length_fields!(bam_record::BAM.Record, record::SAM.Record, header)
+
+    cigar_ops = ('M','I','D','N','S','H','P','=','X')
+    n_cigar_op = sum(x ∈ cigar_ops for x in SAM.cigar(record); init = 0)
+
+    if n_cigar_op > typemax(UInt16) 
+        error("Cigar contains more than $(typemax(UInt16)) operations. This is currently unsupported. See section 4.2.2 of SAM/BAM specifications.")
+    end
```
This check here is unnecessary - you can check the length later, when parsing the CIGAR. This way, you don;t need to create the CIGAR twice for the same record.

jonathanBieler replied:
True, I removed that.

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    isnothing(idx) && error("Coulnd't find Int type for value $(x)")
+    T = candidates[idx]
+
+    T, type_to_aux_int_fields[T]
```
```suggestion
    T, TYPE_TO_AUX_INT_FIELDS[T]
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    cigar = Vector{UInt8}(undef, length(ops)*4)
+    k = 1
+    for c in ops
+        cigar_32 = [UInt32(0) + op_to_code[c.op] + (UInt32(c.L) << 4)]
```
```suggestion
        cigar_32 = [UInt32(0) + OPT_TO_CODE[c.op] + (UInt32(c.L) << 4)]
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    # 1  2  3
+    # AT GG T0
+
+    for i in 1:Nbam
+        k = 2i -1
+        val = BioSequences.encode(DNAAlphabet{4}(), samseq[k]) |> UInt8
+        bamseq[i] += val <<4
+        
+        k = 2i
+        k > Nsam && break
+        val = BioSequences.encode(DNAAlphabet{4}(), samseq[k]) |> UInt8
+        bamseq[i] += val
+    end
```
BioSequences.encode is an internal symbol. Relying on it is risky.
Also, why not loop over `1:Nsam` instead of `1:Nbam`? Then you know for sure that you don't need to break the loop:
```
for (i, sym) in enumerate(samseq)
    val =  BioSequences.encode(DNAAlphabet{4}(), samseq[k]) |> UInt8
    bamseq[div(i - 1, 2)] |= val << (4 * isodd(i))
end
```

jonathanBieler replied:
Done, with div(i + 1, 2)

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+        val = reinterpret(UInt8, [T(val)])
+
+        return vcat(key, typ, val)
+    end
+    if typ == UInt8('Z')
```
```suggestion
    elseif typ == UInt8('Z')
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    if typ == UInt8('Z')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val, 0x00)
+    end
+    if typ == UInt8('A')
```
```suggestion
    elseif typ == UInt8('A')
```

jakobnissen commented on src/convert.jl:
```diff
@@ -0,0 +1,228 @@
+    if typ == UInt8('A')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val)
+    end
+
+    error("Unsupported tag type $(Char(typ))")
```
```suggestion
    else
        error("Unsupported tag type $(Char(typ))")
    end
```
Elseif is more efficient (sometimes)

jakobnissen reviewed (changes requested):
Overall, looks good, with a few comments.
The code allocates a lot, and I think it needs an API to convert a lot of records in bulk, sharing the same header. This might limit this particular implementation.
However, the API is the most important part. It can always be optimised later.

jakobnissen commented:
One more comment, @jonathanBieler - I believe it might be better if you implemented `SAM.Record(::BAM.Record)` instead of a method of `Base.convert`. The reason is that `convert` can be called implicitly, without the programmer's knowledge. This can lead to some problems when conversions happen unexpectedly - implicit copies, degraded performance and more.

jonathanBieler commented:
I think I've addressed most of the issues. Thinking about performances I think the most common use case will be to write BAM records to a file, so maybe it would be possible to add a write method to `BAM.Writer` to write directly `SAM.Record`'s and avoid allocations.

kescobo commented:
I like that plan, but agree with @jakobnissen that getting the API right is the most important part of this PR. Performance can be fixed later, I think. 

@jakobnissen I'm going to let you pull the trigger on merge, but one other thing that could be included here is bumping to `v0.4.1` so that we can tag a release as soon as it's in.

CiaranOMara commented:
Bump the version on a release branch.

kescobo commented:
Ah, my bad - I didn't see that's how this repo is organized. Makes sense to me!

jonathanBieler commented on src/convert.jl:
```diff
@@ -0,0 +1,230 @@
+end
+
+""" See section 5.3 of specifications."""
+function reg2bin(start::Int, stop::Int)
+    stop -= 1
+    (start >> 14) == (stop >> 14) && return ((1 << 15) - 1) ÷ 7 + (start >> 14)
+    (start >> 17) == (stop >> 17) && return ((1 << 12) - 1) ÷ 7 + (start >> 17)
+    (start >> 20) == (stop >> 20) && return ((1 << 9)  - 1) ÷ 7 + (start >> 20)
+    (start >> 23) == (stop >> 23) && return ((1 << 6)  - 1) ÷ 7 + (start >> 23)
+    (start >> 26) == (stop >> 26) && return ((1 << 3)  - 1) ÷ 7 + (start >> 26)
+    return 0
+end
```
Since XAM depends on Indexes and both are based on the SAM specs it would makes sense to use the one in Indexes. But there's a difference in semantics, the one in Indexes, `reg2bins` returns an array on bins while `reg2bin` returns a single bin. I don't understand this enough to make a decision (the code of reg2bin is straight copy-paste from the specs).

CiaranOMara replied:
Apologies, `reg2bin` is certainly needed here for the conversion. The other `reg2bins` is not required for the conversion.

jakobnissen reviewed (approved):
Looks good from my PoV (once @CiaranOMara's comment has been addressed - which ever way you prefer)
